### DISCUSSION TITLE
access trame app externally with token

### DISCUSSION THREAD

**Discussion by bigfooted** [Kitware/trame#262] [Q&A] [answered]
Hi, Is there a way to have trame work with a token or a username/password like a jupyter server? ?

**Comment by jourdain** [comment]
Ideally we'll have examples using Keycloak at some point but something very simple can be done to have a first layout with a form and based on the answer execute the actual app layout with the app. 

Another approach could be to have some basic auth from a proxy like apache.

HTH

**Reply by bigfooted** [reply]
Great thanks! I am not so familiar with what's out there regarding standard authentication software, but keycloak sounds like a good long term solution. Is [trame-keycloak](https://github.com/Kitware/trame-keycloak) a good starting point?

**Reply by jourdain** [reply, marked as answer]
It will be, but I'm not sure it is ready yet. I did not test it yet. Had to work on something else. But it will be once I'm back at it.

One thing that I forgot to mention. Every trame application does use a token but we have a default one which does not make it obvious. 

Are you trying to secure a multi-user setup or a single server/user instance? 

For single server instance just provide your token via `--authKey MySuperSecretToken` and then in the url, you will need to provide `?secret=MySuperSecretToken`.

The same technic can be done with the multi-user setup where you can fix the `--authKey` rather than using the generated one from the launcher to prevent cross session spying.

**Reply by bigfooted** [reply]
Thanks, the authKey keyword works great. At the moment single user is sufficient, but my goal is to have the application available on a public server for multiple users who can all work individually on their own uploaded files, visualize it etc.  
I don't understand how to use this with 'the multi-user setup', what needs to change to make this a multi-user setup? 
I have looked at the Deployment section in the docs, but this is not describing what I want (I think).

**Reply by jourdain** [reply]
So for the multi-user setup the ideal path will be a proper user authentication.

But if the `?secret=abc` approach could be enough, you can tweak the launcher configuration in a way that you provide a fixed `--authKey` so you will only share the secret to the users you want and those who don't know it, won't be able to use the service.

The launcher is just issuing command line to automatically start a new process per users. With the trame docker, we've streamlined that setup to make it "automatic" but you can revert back to "manual" by providing the full command line yourself like we do [here with ParaView](https://github.com/Kitware/paraview-visualizer/blob/master/docker/setup/apps.yml#L4-L18).

**Reply by bigfooted** [reply]
Great thanks! This works for now, and I'll look into keycloak in the couple of weeks.
